### ISSUE TITLE
Add a mechanism to interrupt `Reader::read_line` without user input.

### ISSUE THREAD

**Issue by Lymia** [murarth/linefeed#42]
Currently only user input, or an operating system specific signal can interrupt a running `Reader::read_line` call.

This is not sufficient for concurrent applications that may need to terminate due to reasons other than user input on terminal.

**Comment by murarth** [comment]
Hm. I don't know about this. It seems like bad form to terminate an application while a user may still be typing input. Would it not be possible to set a flag (outside of linefeed) and terminate after the next `read_line` call completes?

**Comment by Lymia** [comment]
The application is basically a background task (a Discord bot), and the console is used for dispatching administrative commands. If an admin uses the shutdown command from Discord rather than terminal, there isn't even a guarantee anyone's connected to the terminal at the moment, much less typing in it.

**Comment by murarth** [comment]
Wouldn't it suffice, then, to print a message to the terminal indicating that the bot has shut down and wait for the user to exit the terminal client?

**Comment by Lymia** [comment]
That would interfere with mechanisms to automatically restart the bot/etc.

And, well... why? It's not an interactive application. For the shutdown command to not actually stop the process, and instead wait for user input sounds completely inane for this use case.

**Comment by murarth** [comment]
I feel like it's not a "clean" exit to interrupt `read_line` and if your application doesn't require a "clean" exit, there are alternatives, in `std::process::exit` or putting `read_line` in a background thread.

**Comment by Lymia** [comment]
I don't understand how forcefully terminating the process with `process::exit` is cleaner than letting all control flow return, by interrupting `read_line`. I understand that it may not be a clean exit for the applications you wrote `linefeed` for, but for the ones I'm using it for, it indeed is far cleaner than anything even resembling waiting for user input to terminate, or calling `exit`. The perfect shouldn't be the enemy of the good.

This is a separate issue, but on Linux, I had issues catching `^C` when running linefeed in a background thread. And regardless, the component that calls read_line holds references to `Arc`s that I would prefer be dropped before the program terminates.

**Comment by murarth** [comment]
I'm not saying `exit` is cleaner. I'm saying that, comparatively, `exit` seems no less clean than interrupting `read_line`, at least on the user interface front.

Perhaps the `read_line` thread could hold weak `Arc` references?

**Comment by Lymia** [comment]
It isn't cleaner on the user interface front, no, but it is far cleaner as far as the actual execution path. I would much rather quit the terminal loop through normal control flow than `exit`, or anything similar.

It will take a bit of work to make sure that everything else exits cleanly, before calling `exit` to quit out just one thread with `read_line`, and making sure everything associated with it is dropped through `Weak`s (for example, the database connection manager.)

... all this, because one thread is refusing to exit for any reason except user input.

Even given these alternative solutions, my evaluation of this is still that the best solution would be to interrupt `read_line` somehow. I don't understand why this is such a point of contention?

**Comment by murarth** [comment]
I've given this issue some thought and I've realized a solution that I think will be mutually agreeable.

The solution I had initially imagined, which I did not want to implement, was to wake `read_line` from its `wait_for_input` call (which currently waits indefinitely) every second or so, to check for a shutdown condition (something like an `AtomicBool`), regardless of whether an application is using this shutdown condition feature.

The more flexible solution that I realized is instead to expose another method, called something like `read_line_step`, which would accept an `Option<Duration>` value. `read_line_step` would wait only this duration for new input, process any new input that becomes available, and return. It would not loop. Thus, an application that wishes to check for a shutdown condition can do so (in whatever way that application desires) and an application that wishes to block indefinitely on new input can do so without waking the main loop unnecessarily.

I will implement this flexible solution shortly and let you know when it's complete.
